adl: fix animations with delay (block animations), refactoring and unittests

obligaron opened the pull request:
This is the successor of #1468

Content:

- Adds correct logic, if delay is specified (fixes fast blocking bug)
- Small refactoring of Calculation, so that the calculation is only done when `NewPlrAnim` is called and not needed every time we render. This also reduces the number of variables we need in `PlayerStruct`.
- Extra comments that explain the logic further
- UnitTests for the Animation Distribution Logic

Notes:

- I wasn't sure if the new flags enum "AnimationFlags" should be `enum class` or normal `enum`.
- If #1507 is commited first, I will add __attribute__((flag_enum)) to the new enum.
- I named the test `AnimationInformation` cause this would be the name of the class when I refactor the logic out of player.h/cpp. if you have a better name, I will change. Other names I had in mind ("AnimationDistribution", "AnimationLogic", "Animation").

Please guide me if something is wrong. Thanks 😊

AJenbo commented:
@obligaron there is no reason to rebase unless you have a conflict :)

obligaron commented:
Sorry didn't know you always get a notification ;-)

AJenbo commented:
No need to rebase unless there are conflicts :)

julealgon reviewed (changes requested):
This is a huge change. Can it be split into smaller refactorings?

In particular, I'm not a fan of refactoring with behavior changes in the same commit. The PR title imples that's the case.

obligaron commented:
@julealgon I thought the same as I wrote the the comment. 
The reason I did it this way, was because the Bugfix wasn't possible without the main-changes introduced with the refactoring. At least it's harder to understand.

- I had to rename variables, cause they changed from representing a Frame to a GameTick.
- I avoided another Member in PlayerStruct, cause this makes it larger and harder to understand.
- ProcessAnimationPending was before and seperated passed as both a skipped Frame and reduce of `_pAnimGameTicksSinceSequenceStarted`. This is not correct in case of specified Animation Delay.

## Main change: Calculation of GameTickModifier

Before:
```
int animationMaxGameTickets = relevantAnimationLength;
if (pPlayer->_pAnimDelay > 1)
	animationMaxGameTickets = (relevantAnimationLength * pPlayer->_pAnimDelay);
float gameTickModifier = (float)animationMaxGameTickets / (float)(relevantAnimationLength - pPlayer->_pAnimNumSkippedFrames);
```

After:
```
int gameTicksPerFrame = (Delay + 1);
int relevantAnimationGameTicksForDistribution = relevantAnimationFramesForDistribution * gameTicksPerFrame;
int relevantAnimationGameTicksWithSkipping = relevantAnimationGameTicksForDistribution - (numSkippedFrames * gameTicksPerFrame);
if (flags & AnimationFlags::ProcessAnimationPending) {
	relevantAnimationGameTicksWithSkipping -= 1;
	plr[pnum]._pAnimGameTicksSinceSequenceStarted = -1;
}
if (flags & AnimationFlags::SkipsDelayOfLastFrame) {
	relevantAnimationGameTicksWithSkipping -= Delay;
}
float gameTickModifier = (float)relevantAnimationGameTicksForDistribution / (float)relevantAnimationGameTicksWithSkipping;
gameTickModifier /= gameTicksPerFrame;
```

A possible split would be:

- Move the original logic to NewPlrAnim and introduce _pAnimGameTickModifier
- Rename _pAnimStopDistributingAfterFrame
- Introduce AnimationFlags::ProcessAnimationPending and stop passing skipped Frames cause of ProcessAnimationPending as skipped Frames
- Change Logic from Frames to GameTicks
- Introduce AnimationFlags::SkipsDelayOfLastFrame and handle it
- include UnitTests

I can give it a try if you want. :relaxed:

If I first wrote the changes this outline wasn't that clear. :flushed:

Sorry if I had burden everyone with extra work. :disappointed_relieved:

julealgon commented:
> I can give it a try if you want. ☺️

If you can isolate smaller, independent changes, that can be done as preliminary refactoring, _even if the final behavior depends on them_, I'd say go for it.

I've been posting this a lot since it makes so much sense to me but here it goes:
> **_make the change easy, then make the easy change_**

The former implies adding preliminary refactoring necessary to make the behavior change as simple and straightforward as possible.

obligaron commented:
I was just a little surprised from your definition of huge. :wink:
I tried to count the changed lines git displayed and skipped every line that changed only cause of comments blanks or }.

This was the result:

Added in | Total | Without Comments, Blanks and }
-- | -- | --
player.h | 16 | 6
player.cpp | 86 | 40
Sum | 102 | **46**


Deleted in | Total | Without Comments, Blanks and }
-- | -- | --
player.h | 6 | 2
player.cpp | 39 | 30
Sum | 45 | **32**

This includes only moved lines. Imho this isn't that huge and I have seen this scale of changed lines in other commits. :thinking:

But I don't want to discuss style, preferences or approach.
I want to contribute to a nice project and I want to learn new stuff (split a commit in an existing pr). :relaxed:
If somethings goes wrong :anguished: I will have to recreate the pr. If this happens sorry in advance.

Thanks for being patient with me. :slightly_smiling_face:

obligaron commented:
As requested the commit is now split into smaller commits.
@AJenbo and @julealgon could you take a look? 🙂

julealgon reviewed (commented):
I feel like the animation system just got insanely more complex lately and I don't feel qualified to review this properly anymore.

I'll leave it to others who have been keeping track of these changes.

On a first glance, they look hackish to me, however.

AJenbo commented on Source/player.cpp:
```diff
@@ -484,9 +484,27 @@ void NewPlrAnim(int pnum, BYTE *Peq, int numFrames, int Delay, int width, int nu
 	plr[pnum]._pAnimGameTicksSinceSequenceStarted = processAnimationPending ? -1 : 0;
-	plr[pnum]._pAnimStopDistributingAfterFrame = stopDistributingAfterFrame;
+	plr[pnum]._pAnimRelevantAnimationFramesForDistributen = 0;
+	plr[pnum]._pAnimGameTickModifier = 0.0f;
```
I believe this is the planned next step in terms of extending this to also work for monsters etc.
(At the very least diablo in Hellfire has a sped up animation)

obligaron replied:
As mentioned in #1468 this would be the next step after this pr.
I try to do small pr's so they are easier to review.

julealgon replied:
> As mentioned in #1468 this would be the next step after this pr.

You should consider reverting the order. Doing cleanup PRs first and then making a change results in simpler code than doing it in the opposite order.

obligaron replied:
In this case the logic wouldn't be simpler. It would be simple shifted to another class.
okay this also reduces the complexity a little. :grin:

but you should consider I didn't plan out to everything at the beginning.
And I think that's normal in agile software development.
As everything else the Animation Distrubtion Logic advances step by step.

But I will try to do the refactoring first if it's foreseeable how the layout of new code will be or what will be needed. 🙂

glebm commented on Source/player.cpp:
```diff
@@ -484,19 +484,32 @@ void NewPlrAnim(int pnum, BYTE *Peq, int numFrames, int Delay, int width, int nu
+			relevantAnimationLength = distributeFramesBeforeFrame - 1;
 		}
+
+		if (flags & AnimationFlags::ProcessAnimationPending) {
```
A flag is always a number with a single `1` bit. So if it's nonzero, it's always that flag.

obligaron replied:
You both have a point.

Consider this enum:
```
enum Countries {
	USA = 1 << 0,
	Canada = 1 << 1,
	NorthAmerica = USA | Canada,
};
```
If we support something like this, @julealgon comment is totally correct.

But as @glebm said currently it's only one single bit.

As I wrote at the pr description. I hesitated if I use `enum `or `enum class`.
I searched and found posts like [this](https://wiggling-bits.net/using-enum-classes-as-type-safe-bitmasks/) that would be imho a better solution. But I didn't want to introduce these new concepts for a simple enum-flag. But if preferred that's clearly possible.
Anyway I also miss a simple `HasFlag` Method.
I'm no c++-Expert. So I finally applied the same style as the surrounding code.

@AJenbo Do your have a opinion?

julealgon replied:
My comment probably stemmed from the fact that I'm not used to having "non-boolean values" used directly as booleans on `if`s. So using an entire bitflag as the if condition seemed very weird to me, but I guess it is acceptable in C++.

But yeah, this crashes badly if the flag is a flag combination.

obligaron replied:
I looked again at the [Code-Style](https://github.com/diasurgical/devilutionX/wiki/Code-Style#conditionschecks-esp-zerofalsy-checks). 
It doesn't state enums explicitly. But the following statement leads me to simply check the result of the &-operator.
> Reasoning: Comparing with the literals of the right type makes the intent of the check clearer in many cases.

glebm commented on Source/player.cpp:
```diff
@@ -514,6 +514,26 @@ void NewPlrAnim(int pnum, BYTE *Peq, int numFrames, int Delay, int width, Animat
 			plr[pnum]._pAnimGameTicksSinceSequenceStarted = -1;
 		}
 
+		if (flags & AnimationFlags::SkipsDelayOfLastFrame) {
```
See reply there

obligaron commented on Source/player.h:
```diff
@@ -179,9 +179,9 @@ struct PlayerStruct {
-	Sint32 _pAnimNumSkippedFrames;              // Number of Frames that will be skipped (for example with modifier "faster attack")
-	Sint32 _pAnimGameTicksSinceSequenceStarted; // Number of GameTicks after the current animation sequence started
-	Sint32 _pAnimStopDistributingAfterFrame;    // Distribute the NumSkippedFrames only before this frame
+	float _pAnimGameTickModifier;                       // specifies how many animations-fractions are displayed between two gameticks. this can be > 0, if animations are skipped or < 0 if the same animation is shown in multiple times (delay specified).
```
Sure, was just keeping the same comment style as the surrounding code. Fixed 🙂

obligaron commented on Source/player.h:
```diff
@@ -378,6 +378,15 @@ void LoadPlrGFX(int pnum, player_graphic gfxflag);
+/**
+ * @brief Specifies what special logics are applied for a Animation
+ */
+enum AnimationFlags : uint8_t {
```
I also thought about `AnimationDistributionFlags` cause it's only used for this.
What do you think?

julealgon replied:
`AnimationFrameDistributionFlags`, or even just `FrameDistributionFlags` then maybe?

obligaron replied:
I avoided Frame, cause the enum also have special casing for delay and all Methods and Fields are called something with Animation (NewPlrAnim. _pAnimLen, ...).
But I renamed to `AnimationDistributionFlags` with latest commit.
I hope that's okay.

obligaron commented on Source/player.cpp:
```diff
@@ -489,33 +489,35 @@ void NewPlrAnim(int pnum, BYTE *Peq, int numFrames, int Delay, int width, Animat
+			relevantAnimationFramesForDistribution = distributeFramesBeforeFrame - 1;
 		}
 
+		int gameTicksPerFrame = (Delay + 1);                                                                        // How many GameTicks are needed to advance one Animation Frame
+		int relevantAnimationGameTicksForDistribution = relevantAnimationFramesForDistribution * gameTicksPerFrame; // GameTicks that will be adjusted for the skipped Frames/GameTicks
+
+		int relevantAnimationGameTicksWithSkipping = relevantAnimationGameTicksForDistribution - (numSkippedFrames * gameTicksPerFrame); // How many GameTicks will the Animation be really shown (skipped Frames and GameTicks removed)
```
I did this cause then it's more clear that the comment only applies to this specific line.
But I don't mind changing it if this helps.
Question: Only here or should i avoid adding comments at the same line at all occurrences?

Example:
```
// this can happen if we are at the last frame and the next game tick is due (nthread_GetProgressToNextGameTick returns 1.0f)
if (absolutAnimationFrame > relevantAnimationFrames) {
```

	
`	if (absolutAnimationFrame > relevantAnimationFrames) {  // this can happen if we are at the last frame and the next game tick is due (nthread_GetProgressToNextGameTick returns 1.0f)`

julealgon replied:
> Question: Only here or should i avoid adding comments at the same line at all occurrences?

IMHO, always avoid it. Place them above, and leave a blank line above the comment itself to make it clear it is commenting the block below it.

obligaron replied:
Changed it according to your guidance.
And I think in most cases it's more clear.
Thank you 🙂

obligaron commented on Source/player.cpp:
```diff
@@ -3679,30 +3697,20 @@ void ProcessPlayerAnimation(int pnum)
-	float gameTickModifier = (float)animationMaxGameTickets / (float)(relevantAnimationLength - pPlayer->_pAnimNumSkippedFrames); // if we skipped Frames we need to expand the GameTicks to make one GameTick for this Animation "faster"
-	int absolutAnimationFrame = 1 + (int)(totalGameTicksForCurrentAnimationSequence * gameTickModifier); // 1 added for rounding reasons. float to int cast always truncate.
+
+	int absolutAnimationFrame = 1 + (int)(totalGameTicksForCurrentAnimationSequence * pPlayer->_pAnimGameTickModifier); // 1 added for rounding reasons. float to int cast always truncate.
```
fixed. thanks 🙂

obligaron commented on Source/player.cpp:
```diff
@@ -484,9 +484,27 @@ void NewPlrAnim(int pnum, BYTE *Peq, int numFrames, int Delay, int width, int nu
-	plr[pnum]._pAnimNumSkippedFrames = numSkippedFrames;
 	plr[pnum]._pAnimGameTicksSinceSequenceStarted = processAnimationPending ? -1 : 0;
-	plr[pnum]._pAnimStopDistributingAfterFrame = stopDistributingAfterFrame;
+	plr[pnum]._pAnimRelevantAnimationFramesForDistributen = 0;
```
fixed. thanks 🙂

yuripourre commented on Source/player.cpp:
```diff
@@ -484,9 +484,62 @@ void NewPlrAnim(int pnum, BYTE *Peq, int numFrames, int Delay, int width, int nu
+	plr[pnum]._pAnimGameTickModifier = 0.0f;
+
+	if (numSkippedFrames != 0 || flags != AnimationFlags::None) {
+		int relevantAnimationFramesForDistributing = numFrames; // Animation Frames that will be adjusted for the skipped Frames/GameTicks
```
Hey, @obligaron I am trying to understand the logic, this variable represents the remaining frames, right?

obligaron replied:
Yes you could say this. But let me clarify more:
At this point we initialize the Animation. We must know how many Frames the Animation is in total. Cause we want to display every bit of nice content (every Frame) in it.
Normally this should be total frames (`numFrames`). But in an Player-Attack Animation this would result in some glitches.
The Sword Attack Animation for Warrior has 16 Frames.
In the 9 Frames the enemy gets hit and the hit-sound is played.
So the hit and the hit-sound happens after 56% (9/16) of the Animation is finished.

If we have a skipping of 4 frames (fastest attack modifier), this means we end up with 12 shown Frames and the hit happens at frame 5 (cause skipping is always applied before the hit animation).
The hit and the hit-sound happens now after 41% of the Animation is finished.

If we would now distribute all frames equal, the hit and the hit-sound would be out of sync.
I hoped this helps.
If you have a suggestion how the comments can be improved, please let me know. 🙂

yuripourre replied:
Thanks @obligaron! I understood the code after reading a couple times. But needed to confirm if I got it right.

It's hard to suggest better comments you did a great job.

Maybe I would renaming `relevantAnimationFramesForDistributing` to `remainingFrames`. Is hard to even suggest better variable names too.

obligaron replied:
Naming is always hard :smiley: but really important.
I would stick with `relevantAnimationFramesForDistributing` cause my brain thinks `remainingFrames` implies some sort of progress. And at this point that's not the case.

Anyway thank you for your comment. It was really motivating. :smile:

obligaron commented:
> 
> 
> I feel like the animation system just got insanely more complex lately and I don't feel qualified to review this properly anymore.
> 

It sure got more complex. The main reason is that we decoupled GameLogic and Animations. And this is a new concept with new complexity.

As far as I understood this was a long wished improvement. And it really can make a visible change.

The question is the currently approach good? **Can we do better?**

# Restrictions

The Implementation has to regard some imposed burdens. Perhaps some of them are discussable.

- The Game Balance shouldn't change.
- Compatibility for save games should be maintained.

# Challenges

- How much of the existing GameLogic must be adjusted. What do we want? Can this stay mainly in the rendering?
- How can we avoid that we regress the animation logic in later changes?
- What happens if the Rendering fails?

# Alternatives

## 1. Do Nothing

We simply could say, this adds to much complexity and work. Revert the newly added Animation Distribution Logic and stay with a simpler and easier to understand solution. That's totally possible.

## 2. Let the Animation know how many GameTicks the Animation will be active

This is the currently approach. We tell the rendering how many GameTicks the Animation is active. This way it can distribute **all** Animation Frames between the starting and the ending GameTick. That's also the reason why we need to know how long it is to the next GameTick, so we can distribute accordingly.

The downside is, you have to handle the skipping logic and other details at two places. In the original GameLogic (for example in `PM_DoAttack`) and for the new Animation Distribution Logic (for example in `StartAttack`). That's not nice and clean.

But perhaps this can be changed later. We could use the Information passed at the Start of the Animation also in `ProcessPlayerAnimation` and remove the logic entirely from original GameLogic (for example in `PM_DoAttack`).
I didn't start with that, cause this would make the changes (prs) much bigger and disruptive. Also it's a discussable topic, cause the question arises how much we want to stray from the original design.
So I started with smaller changes. But I hope we can look at this in the future. 🙂

## 3. Some approach that I didn't figure out

It's clearly possible that there can be other better possible implementations. That's why we discuss this. 🙂 

# Conclusion

I tried to start with smaller changes, that wouldn't affect too much of the GameLogic.
The complexity is the reason why I wrote so [many comments](https://github.com/diasurgical/devilutionX/pull/1539#discussion_r615139002). I wanted that it's possible to understand the logic and the reasons why some statements are introduced. If they are not clear or not enough let me know so we can improve them together. 🙂

obligaron commented on test/animationinformation_test.cpp:
```diff
@@ -0,0 +1,371 @@
+	        new GameTickData(16, 0),
+	        new RenderingData(0.6f, 16),
+	        // Animation stopped cause PM_DoAttack would stop the Animation "if (plr[pnum]._pAnimFrame == plr[pnum]._pAFrames) {"
+	    });
```
I added a comment. Does this help?
Have you other suggestions what can be changed to improve the unittest?

julealgon replied:
Comments to explain code should be avoided as much as possible. The problem I had here is that the code itself is really hard to reason about, and thus the code itself should be improved instead.

> Have you other suggestions what can be changed to improve the unittest?

Not really no... as I said this change has got a bit out of hand for me, and it is tough to suggest anything if I don't clearly understand everything going on.

I'm just giving feedback as someone "looking from a distance", basically, so take it with a grain of salt.

obligaron replied:
I would also say it's best if the code can be understand without comments. that's also why good naming is really important.
But with increasing complexity it's not always possible to avoid comments.
I wanted to leave something in the Sourcecode that other people could help in this regard.

That's also why we need to read documentations of our used language or framework.
If you don't mind I would close this discussion as I would need input how I can help with this matter.
The bigger picture and decision decision we have to take is described in my last [comment](https://github.com/diasurgical/devilutionX/pull/1539#issuecomment-821762887).

But I'm open to any future feedback, as I want a good solution for this project.

julealgon replied:
> If you don't mind I would close this discussion as I would need input how I can help with this matter.

I don't mind at all. It was added to see if you could come up with an alternative yourself.

Maybe other folks can provide more guidance than me on this particular one.

As i said, i don't feel qualified to properly review a good portion of this code.

obligaron commented on Source/player.cpp:
```diff
@@ -484,19 +484,32 @@ void NewPlrAnim(int pnum, BYTE *Peq, int numFrames, int Delay, int width, int nu
+			// The Animation Distribution Logic needs to account how many GameTicks passed since the Animation started.
+			// Because ProcessAnimation will increase this later (in same GameTick as NewPlrAnim), we correct this upfront.
+			// This also means Rendering should never hapen with _pAnimGameTicksSinceSequenceStarted < 0.
+			plr[pnum]._pAnimGameTicksSinceSequenceStarted = -1;
```
See comment in other [conversation](https://github.com/diasurgical/devilutionX/pull/1539#discussion_r615294730).

As state there: I'm open to any future feedback, as I want a good solution for this project.

obligaron commented on Source/player.cpp:
```diff
@@ -484,9 +484,27 @@ void NewPlrAnim(int pnum, BYTE *Peq, int numFrames, int Delay, int width, int nu
+		int animationMaxGameTickets = relevantAnimationLength;
+		if (Delay > 1)
+			animationMaxGameTickets = (relevantAnimationLength * Delay);
+		float gameTickModifier = (float)animationMaxGameTickets / (float)(relevantAnimationLength - numSkippedFrames); // if we skipped Frames we need to expand the GameTicks to make one GameTick for this Animation "faster"
```
See comment in other [conversation](https://github.com/diasurgical/devilutionX/pull/1539#discussion_r615294730).

As state there: I'm open to any future feedback, as I want a good solution for this project.

AJenbo commented:
Sorry for being slow to review this, I read text much slower then code :sweat_smile:

obligaron commented:
No Problem. This pr shouldn't impose stress to anyone. 🙂 
I'm just eager to improve the logic even further (refactoring) and expand it to other types (monsters, ...). :upside_down_face:

obligaron commented:
I discovered a issue with walking animations in town.
For walk the normal/previous animation logic is now used.

I have a bugfix at hand, but didn't want to make the pr larger.

AJenbo commented on Source/player.h:
```diff
@@ -397,6 +406,22 @@ void LoadPlrGFX(int pnum, player_graphic gfxflag);
+	/*
+	* @brief Delay of last Frame is ignored (for example, because only Frame and not delay is checked in game_logic)
+	*/
+	SkipsDelayOfLastFrame = 1 << 1,
```
Why specify these as bit flag values? You aren't using combinations anywhere as far as I can see?

obligaron replied:
After looking at `monsterdata` I saw some animations with delay that aren't the standing animation. So I thought there will be animations that needs a combination.
And I dislike a method calls like `NewAnimation(..., true, false, true, false, true)`. :smile:

But as you said, currently they are not needed.
If you want I can convert them to `enum class` and introduce the flag-logic later if we encounter the specific animation. 🙂

AJenbo replied:
I world prefer that :)

obligaron replied:
done 🙂

AJenbo reviewed (dismissed):
Finally took the time to review this, all the test makes it really daunting to look at :D

For the most part, this looks good, but I'm not sure if there really is a need for having the enums works as bit flags.
